ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala

```diff
@@ -79,60 +84,85 @@ class ContractsService(
-        .collect {
-          case e @ -\/(_) => e
-          case a @ \/-(ac) if predicate(ac) => a
-        }
-        .runWith(Sink.headOption): Future[Option[Error \/ domain.ActiveContract[LfValue]]]
+  private object SearchInMemory extends Search {
+    type LfV = LfValue
+    override val lfvToJsValue = SearchValueFormat(lfValueToJsValue)
 
-      result <- lookupResult(errorOrAc)
+    override def findByContractKey(
+        ctx: SearchContext[Id, Option],
+        contractKey: LfValue,
+    ): Future[Option[domain.ActiveContract[LfValue]]] = {
+      import ctx.{jwt, parties, templateIds => templateId}
+      for {
+        resolvedTemplateId <- toFuture(resolveTemplateId(templateId)): Future[
+          TemplateId.RequiredPkg]
 
-    } yield result
+        predicate = domain.ActiveContract.matchesKey(contractKey) _
 
-  def findByContractId(
-      jwt: Jwt,
-      parties: OneAnd[Set, lar.Party],
-      templateId: Option[domain.TemplateId.OptionalPkg],
-      contractId: domain.ContractId,
-  ): Future[Option[domain.ActiveContract[LfValue]]] =
-    for {
-
-      resolvedTemplateIds <- templateId.cata(
-        x => toFuture(resolveTemplateId(x).map(Set(_))),
-        Future.successful(allTemplateIds()),
-      ): Future[Set[domain.TemplateId.RequiredPkg]]
-
-      errorOrAc <- searchInMemory(jwt, parties, resolvedTemplateIds, Map.empty)
-        .collect {
-          case e @ -\/(_) => e
-          case a @ \/-(ac) if isContractId(contractId)(ac) => a
-        }
-        .runWith(Sink.headOption): Future[Option[Error \/ domain.ActiveContract[LfValue]]]
+        errorOrAc <- searchInMemoryOneTpId(jwt, parties, resolvedTemplateId, predicate)
+          .runWith(Sink.headOption): Future[Option[Error \/ domain.ActiveContract[LfValue]]]
+
+        result <- lookupResult(errorOrAc)
+
+      } yield result
+    }
+
+    override def findByContractId(
+        ctx: SearchContext[Option, Option],
+        contractId: domain.ContractId,
+    ): Future[Option[domain.ActiveContract[LfValue]]] = {
+      import ctx.{jwt, parties, templateIds => templateId}
+      for {
 
-      result <- lookupResult(errorOrAc)
+        resolvedTemplateIds <- templateId.cata(
+          x => toFuture(resolveTemplateId(x).map(Set(_))),
+          Future.successful(allTemplateIds()),
+        ): Future[Set[domain.TemplateId.RequiredPkg]]
 
-    } yield result
+        errorOrAc <- searchInMemory(
+          jwt,
+          parties,
+          resolvedTemplateIds,
+          InMemoryQuery.Filter(isContractId(contractId)))
+          .runWith(Sink.headOption): Future[Option[Error \/ domain.ActiveContract[LfValue]]]
+
+        result <- lookupResult(errorOrAc)
+
+      } yield result
+    }
+
+    override def search(ctx: SearchContext[Set, Id], queryParams: Map[String, JsValue]) = {
+      import ctx.{jwt, parties, templateIds}
+      searchInMemory(jwt, parties, templateIds, InMemoryQuery.Params(queryParams))
```

Comment:
This didn’t change in this PR but why are we representing queries as `Map[String, JsValue]`. This seems like we are encoding the fact that the first layer is a record? Is that useful given that we lose all structure on the levels below and therefore have to deal with incorrectly typed queries anyway?

Reply:
It's been a while, but as I vaguely recall, certain invariants are easier to enforce if you type this fact. And all else being equal I'm loath to give up a carefully typed premise.